Feature: Offer logout option in auth provider

andrico1234 opened the pull request:
Closes: https://github.com/marmelab/react-admin/issues/5882

Handles additional options to the `checkError`'s `Promise.reject`.


https://user-images.githubusercontent.com/17087167/120549111-05fb0580-c3eb-11eb-944b-9f150da6fa0d.mov

##  Notes: 

- Do we want to rename the `useLogoutIfAccessDenied`, since it can now sometimes not log the user out if access is denied.
- Should we move the redirect building logic out to a reusable function, since it’s very similar to the one used in `useLogout`. 
- Do we also want to add similar logic to the checkAuth function?

andrico1234 commented on packages/ra-core/src/auth/useLogoutIfAccessDenied.spec.tsx:
```diff
@@ -67,11 +79,12 @@ describe('useLogoutIfAccessDenied', () => {
 
     it('should not logout if passed no error', async () => {
-        const { queryByText } = render(
+        const { queryByText } = renderComponent(
```
We can either mock the `useHistory`, or provide a Router. Doing neither results in this: 
![image](https://user-images.githubusercontent.com/17087167/120549511-9cc7c200-c3eb-11eb-9f51-3f9876ee276b.png)

I chose to wrap them all in a router because

a) I prefer to avoid mocking if I don't need to
b) we'd have to wrap the new test in a router regardless, and so we'd be wrangling with both sets logic in the suite

andrico1234 commented:
@djhi @fzaninotto hey folks! wondering if there's any update for this ticket? It's good for a review

djhi commented on docs/Authentication.md:
```diff
@@ -186,6 +186,26 @@ export default {
+    checkError: (error) => {
+        const status = error.status;
+        if (status === 401 || status === 403) {
+            localStorage.removeItem('auth');
```
I think you can remove this line

andrico1234 replied:
removed

djhi commented on packages/ra-core/src/auth/useLogoutIfAccessDenied.ts:
```diff
@@ -76,11 +81,33 @@ const useLogoutIfAccessDenied = (): LogoutIfAccessDenied => {
+                    if (logoutUser) {
+                        logout({}, redirectTo);
+                    } else {
+                        const redirectToParts = redirectTo.split('?');
+                        const newLocation: LocationDescriptorObject = {
+                            pathname: redirectToParts[0],
+                        };
+
+                        if (history.location && history.location.pathname) {
+                            newLocation.state = {
+                                nextPathname: history.location.pathname,
+                                nextSearch: history.location.search,
+                            };
+                        }
+
+                        if (redirectToParts[1]) {
+                            newLocation.search = redirectToParts[1];
+                        }
+                        history.push(redirectTo);
```
We should probably extract this from here and from `useLogout`. @fzaninotto ?

andrico1234 replied:
I thought the same too, but wasn't sure where the right place for it it is. What do you both think?

fzaninotto replied:
Shouldn't this be `history.push(newLocation)`? Otherwise the code above is useless.

Also, setting `nextPathName` and `nextSearch` is currently only useful for the Login page. I don't see any point in having them here. And since we don't need that, we don't need to manipulate the `redirectTo`.

andrico1234 replied:
It's been a few weeks since I've gone over this code, and I'll sanity check just in case, but I believe the `history.push(redirectTo)` is the correct behaviour. In essence, we want to perform the same behaviour but as the `logout({}, redirectTo)`but to circumvent the logout functionality.

andrico1234 replied:
@fzaninotto , does the video in the PR's description clarify things?

andrico1234 replied:
ignore me, you're completely right!

djhi reviewed (changes requested):
Thanks!

andrico1234 commented:
I've resolved the requested changes, @djhi

fzaninotto commented on docs/Authentication.md:
```diff
@@ -491,40 +510,40 @@ const authProvider = {
-| `logout`         | Log a user out                                  |                    |
-| `getIdentity`    | Get the current user identity                   |                    | 
-| `getPermissions` | Get the current user credentials                | `Object` whatever params passed to `usePermissions()` - empty for react-admin default routes |
+| Method           | Usage                                                    | Parameters format                                                                            |
```
please revert these table formatting changes. They make the doc harder to read when not translated into Markdown, and they don't bring any value.

andrico1234 replied:
I think it's related to my prettier config, this happens every time! I'll revert

fzaninotto commented on docs/Authentication.md:
```diff
@@ -491,40 +510,40 @@ const authProvider = {
-| `checkAuth`      | User is authenticated             | `void`          |
-| `logout`         | Auth backend acknowledged logout  | `string | false | void` route to redirect to after logout, defaults to `/login` |
-| `getIdentity`    | Auth backend returned identity    | `{ id: string | number, fullName?: string, avatar?: string }`  | 
+| Method           | Resolve if                        | Response format                                                                                |
```
same

fzaninotto commented on docs/Authentication.md:
```diff
@@ -491,40 +510,40 @@ const authProvider = {
-| Method           | Reject if                                 | Error format |
-| ---------------- | ----------------------------------------- | --------------- |
-| `login`          | Login credentials weren't accepted        | `string | { message?: string }` error message to display |
+| Method           | Reject if                                 | Error format                                                                                                                     |
```
same

fzaninotto commented on packages/ra-core/src/auth/useLogoutIfAccessDenied.spec.tsx:
```diff
@@ -58,6 +60,16 @@ const notify = jest.fn();
 //@ts-expect-error
 useNotify.mockImplementation(() => notify);
 
+function renderComponent(children) {
```
please rename to `renderInRouter`

fzaninotto reviewed (changes requested):
Nice!

andrico1234 commented:
@djhi @fzaninotto 

I'd like to get a Cypress test added for this behaviour, but can't figure out an appropriate way to do it.

Is there a way to monkey patch in a slightly modified authProvider? If not, I could:
- create a really basic RA-app, but that seems like overkill.
- create a new resource that specifically logs the user when clicking through to it (but we'd still need to alter the authProvider)

I added a unit test, but it didn't catch the problem when I called `history.push(redirectTo)` instead of `history.push(newLocation)`.

What do you guys think?

Other than that, i've resolved the rest of the comments

andrico1234 commented:
Note: I added a notification if a user was redirected but not logged out. I added a french translation that was made via google translate, since I don't speak French! Please let me know what you'd like the message to be 😄

fzaninotto commented on packages/ra-core/src/auth/useLogoutIfAccessDenied.ts:
```diff
@@ -76,11 +91,20 @@ const useLogoutIfAccessDenied = (): LogoutIfAccessDenied => {
+                    if (logoutUser) {
+                        logout({}, redirectTo);
+                    } else {
+                        const newLocation: LocationDescriptorObject = {
```
Why do you create a new object? This forbids developers to pass a location object themselves. 

I'd pass the `redirectTo` directly, like so:

```
history.push(redirectTo);
```

andrico1234 replied:
How do you recommend we do this then?

Specify the `redirectTo` object be a type of `LocationDescriptorObject`, and invert control up to the authProvider?

andrico1234 replied:
Ignore the above comment, i had some issue with testing previously. I've pushed your suggestions and it works fine. 

I've tested using a regular path like `/unauthorized` and also adding query params. Are there any other manual test cases you recommend I try?

fzaninotto commented on packages/ra-language-french/src/index.ts:
```diff
@@ -143,6 +143,8 @@ const frenchMessages: TranslationMessages = {
             canceled: 'Action annulée',
             logged_out: 'Votre session a pris fin, veuillez vous reconnecter.',
+            not_authorized:
+                "Vous n'êtes pas autorisé à accéder à cette ressource.",
```
```suggestion
                "Vous n'êtes pas autorisé(e) à accéder à cette ressource.",
```

fzaninotto reviewed (changes requested):
Awesome!

fzaninotto commented:
Thanks a lot!
